Load dense pose partitions. Use them for sensor indices

una-dinosauria opened the pull request:
Implements a way to incorporate partitions into the LogReader.

Basically, the LogReader now gets an extra argument: `partitions: Optional[Iterable[Partition]] = None,` which filters poses and indices and only returns the ones that lie at the intersection of the passed partitions. Eg, setting 

```
partitions=[PreProcessPartition.VALID, GeoPartition.TRAIN],
```
Will only returned the poses/index entries that remain after pre-filtering, and that belong to the TRAIN partition, and 

```
partitions=[PreProcessPartition.VALID, GeoPartition.TEST, QueryBasePartition.QUERY]
```

Will only return poses/index entries that lie in both the TEST and QUERY partitions.

AndreiBarsan commented:
Are there any samples filtered by `VALID` (i.e., which are invalid) which are not already filtered by `TINY`/`MID`/`FULL`?

AndreiBarsan commented on pit30m/data/log_reader.py:
```diff
@@ -143,24 +148,50 @@ def __repr__(self) -> str:
-        if not self.partitions:
-            n_sensor_measurements = len(self.get_cam_geo_index(CamName.MIDDLE_FRONT_WIDE))
-            return np.full(n_sensor_measurements, True)
+        # breakpoint()
```
Nit: consider removing

una-dinosauria replied:
Done.

AndreiBarsan commented on pit30m/data/log_reader.py:
```diff
@@ -143,24 +148,50 @@ def __repr__(self) -> str:
+        combined_indices_mask = np.logical_and.reduce(partition_indices)
+        timestamps = self.raw_pose_data["capture_time"]
+
+        # HACK: Convert the raw poses' timestamps to UNIX to make them comparable to those... everywhere else
```
Could you guard this so we only do it for v1? Indexes v2 and on will use GPS time as discussed ofline.

una-dinosauria replied:
Would you be ok with re-naming the indices to v0 before the dataset is announced? I'd rather not have custom logic for versions that people are not supposed to use.

una-dinosauria replied:
Moved to v2.

AndreiBarsan commented on pit30m/data/log_reader.py:
```diff
@@ -143,24 +148,50 @@ def __repr__(self) -> str:
+        we must index sensor data into these partitions to find out which partitions they belong to.
+
+        Returns:
+            A tuple of boolean arrays. Each array has n_dense_poses entries, and each entry represents membership of the
```
How many elements? If the count will vary based on the context, would a `list` be a better fit?

una-dinosauria replied:
Right, moved it to a list!

AndreiBarsan commented on pit30m/data/log_reader.py:
```diff
@@ -223,23 +254,50 @@ def get_lidar_geo_index(self, sort_by: str = "lidar_time") -> np.ndarray:
+        self,
+        cam_name: CamName = CamName.MIDDLE_FRONT_WIDE,
+        sort_by: str = "img_time",
+        max_delta_s=0.1,
```
Nit: document `max_delta_s`

una-dinosauria replied:
Done!

AndreiBarsan commented on pit30m/data/log_reader.py:
```diff
@@ -223,23 +254,50 @@ def get_lidar_geo_index(self, sort_by: str = "lidar_time") -> np.ndarray:
         index = index[np.argsort(index[sort_by])]
+
+        # Filter by partitions
+        partitions_mask = self.filter_partitions(index["img_time"], max_delta_s)
```
Are you sure this will be fast enough to always run on-demand whenever someone needs to read a log?

una-dinosauria replied:
This is surprisingly fast! And we can always make it faster with a custom implementation taking advantage of the fact that both arrays are sorted. I think the simplicity of being able to drop-in new indices or new partitions is well-worth it.

AndreiBarsan replied:
Sounds awesome! Thanks for confirming!

AndreiBarsan commented on pit30m/data/submap.py:
```diff
@@ -127,6 +130,7 @@ def to_utm(self, map_poses_xyz: np.ndarray, submap_ids: Iterable[UUID], strict:
 
         off_utm = np.array(off_utm)
         result = np.array(map_poses_xyz)
+        breakpoint()
```
debug?

AndreiBarsan commented on pit30m/indexing.py:
```diff
@@ -284,18 +288,79 @@ def fetch_metadata_for_lidar(lidar_uri: str) -> tuple[str, tuple]:
+def associate_np(
+    query_timestamps: np.ndarray, target_timestamps: np.ndarray, max_delta_s: float = 0.5
+) -> Tuple[np.ndarray, np.ndarray]:
+    """Similar associate, but fully vectorized
```
Should this just replace `associate`?

una-dinosauria commented:
Thanks for your comments @AndreiBarsan. It occurred to me that it might be better to make the passed set of partitions explicit when the user asks for an array of poses or an index, since this way we can cache the matching and other heavy objects, and re-use the same log-reader to eg, load both train and validation parts of a dataset. I'll implement this change and ping you again, sorry for the trouble.

una-dinosauria commented:
I've addressed all the comments, and made the changes that I mentioned -- the filtering is now done only when getting the index, and nowhere else. This means that we can re-use the same `LogReader` to fetch different partitions, which is much faster and efficient than creating different `LogReader`s per partition combination.

The notebook showcases how to do this, and plots train/val/test poses from image indices from 50 logs:
<img width="713" alt="image" src="https://github.com/pit30m/pit30m/assets/3733964/df723821-a374-446f-b936-c7bf2f58afc9">

AndreiBarsan commented on pit30m/data/log_reader.py:
```diff
@@ -127,53 +138,116 @@ def __init__(
+
+    def partition_assigments(self, partitions: Optional[Iterable[Partition]] = None) -> List[np.ndarray]:
+        """Fetches partition indices from S3 and converts them to boolean arrays according to the requested partition
+        values. Note that these partitions are wrt the raw pose data @ 100Hz, not the sensor data; therefore, we must
```
```suggestion
        values. Note that these partitions are wrt the raw pose data @ 100Hz, not the sensor data (i.e. not wrt any index); therefore, we must
```

AndreiBarsan commented on pit30m/data/log_reader.py:
```diff
@@ -127,53 +138,116 @@ def __init__(
-        partition_indices = self.partition_assignments
-        combined_indices = np.logical_and.reduce(partition_indices)
-        return combined_indices
+        TODO(julieta) This is a rather bandwidth-heavy operation because to load the partiton timestamps we load the
```
Would you mind creating a GH issue to track this? I can look into it

AndreiBarsan commented on tests/data/log_reader_integration_test.py:
```diff
@@ -18,32 +18,6 @@ def _real_log_reader(real_map: Map) -> LogReader:
 
 
-@fixture(name="real_log_reader_with_test_partition", scope="session")
-def _real_log_reader_test_query(real_map: Map) -> LogReader:
```
is removing this fixture intentional?

AndreiBarsan reviewed (approved):
LGTM with minor comments!
